sstables/generation_type.hh

```diff
@@ -12,43 +12,50 @@
 #include <compare>
 #include <limits>
 #include <iostream>
+#include <variant>
 #include <seastar/core/sstring.hh>
+#include "utils/UUID.hh"
 
 namespace sstables {
 
-class generation_type {
-    int64_t _value;
-public:
+struct generation_type {
+    using impl_type = std::variant<int64_t, utils::UUID>;
+    impl_type value;
+
     generation_type() = delete;
 
-    explicit constexpr generation_type(int64_t value) noexcept: _value(value) {}
-    constexpr int64_t value() const noexcept { return _value; }
+    constexpr generation_type(const generation_type& gen) noexcept = default;
+    constexpr generation_type& operator=(const generation_type& gen) noexcept = default;
+
+    template <typename T>
+    explicit constexpr generation_type(const T& _value) noexcept: value(_value) {}
 
-    constexpr bool operator==(const generation_type& other) const noexcept { return _value == other._value; }
-    constexpr std::strong_ordering operator<=>(const generation_type& other) const noexcept { return _value <=> other._value; }
+    constexpr std::strong_ordering operator<=>(const generation_type& other) const noexcept = default;
 };
 
-constexpr generation_type generation_from_value(int64_t value) {
-    return generation_type{value};
+inline auto generation_value(const generation_type& generation) {
+    assert(false && "unimplemented"); // TODO(JS): implement
+    return 0;
```

Comment:
Keeping the order of declaration (`generation_from_value` before `generation_value`) may reduce the churn in this patch.

Reply:
Also, won't this break bisectibility?